### PR TITLE
Ford documentation: improvements

### PR DESCRIPTION
**Pull request by jvdp1**
Related to #783 by @ZedThree

Proposition:
* Ford: upgrade from `6.1.10` to `7.0.5`
* exclude all BLAS/LAPACK source files from the `ford` processing (2 reasons: 1) too many procedures; 2) no support of both `fypp` and `cpp` by `ford`)
* `stdlib_str2num.fypp`: small fixes for `ford` processing

### CODE REVIEW

**Comment by jvdp1** [comment]
@zedthree it seems that this latest version of `ford` (`7.0.5`) does not generate the links between the specs to the code, e.g., for `call [[stdlib_sorting(module):ord_sort(interface)]]( array[, work, reverse ] )`. Has the notation been changed between `6.1.10` and `7.0.5`?

Also, it seems that it is not possible to exclude files in `6.1.10`. Hence, the need to upgrade `ford` to `7.0.5`.

**Comment by gareth-nx** [comment]
Thanks for this.

IMO it would be good to include documentation for the BLAS/LAPACK interfaces, if not the individual routines for each precision. Otherwise I think it will be very easy for users to become confused about what is available.

I'm not sure whether your pull request keeps the interfaces or not?

**Comment by jvdp1** [comment]
> IMO it would be good to include documentation for the BLAS/LAPACK interfaces, if not the individual routines for each precision. Otherwise I think it will be very easy for users to become confused about what is available.
> 
> I'm not sure whether your pull request keeps the interfaces or not?

A description of BLAS/LAPACK is available in the [specs](https://stdlib.fortran-lang.org/page/specs/stdlib_linalg.html#blas-and-lapack). It was decided that only a link would be provided to the individual LAPACK interfaces.

However, I will try to find a way to process only one of the precisions. It could be indeed useful.

**Review by perazz — APPROVED** [review]

**Comment by perazz** [comment]
These changes look great to me. 
@jvdp1 @gareth-nx @ZedThree: In PR #786 I've addressed the FORD syntax issues, so, FORD can now process all BLAS and LAPACK sources just fine. However, I do confirm that it is a lengthy process (at least on my Mac). 

I think that the best thing to do would be to exclude the sources, and have FORD only parse the interface blocks. However, FORD will fail in this case (because it cannot find the module procedures in the interface)

**Comment by ZedThree** [comment]
@jvdp1 The issue with that particular example in `stdlib_sorting.md` is that it is within backticks:

```markdown
##### Syntax

`call [[stdlib_sorting(module):ord_sort(interface)]]( array[, work, reverse ] )`
```

and in Ford v7 the priority has changed such that backticks now (correctly IMO) are always formatted verbatim, so this needs to be something like:

```markdown
##### Syntax

`call` [[stdlib_sorting(module):ord_sort(interface)]] `( array[, work, reverse ] )`
```

A quick search shows about 270 examples of links within code blocks like this. A minimal change to ford to restore the old behaviour is:

```diff
modified   ford/_markdown.py
@@ -271,7 +271,7 @@ class FordLinkExtension(Extension):
     def extendMarkdown(self, md: MetaMarkdown):  # type: ignore[override]
         project = self.getConfig("project")
         md.inlinePatterns.register(
-            FordLinkProcessor(md, project=project), "ford_links", 174
+            FordLinkProcessor(md, project=project), "ford_links", 194
         )
 
```

but this would really need to be user-configurable, as otherwise it's not possible to write links verbatim (in general, nesting of inline markup is a bit of a nightmare of conflicting syntax, so this will always be a compromise).

If there's a strong desire for this, I can add this option and get it out in 7.0.6

**Comment by jvdp1** [comment]
Thank you @ZedThree for your explanations.

If the following syntax works with Ford v7, I can modify the specs within this PR to support it

> ```
> ##### Syntax
> 
> `call` [[stdlib_sorting(module):ord_sort(interface)]] `( array[, work, reverse ] )`
> ```

> If there's a strong desire for this, I can add this option and get it out in 7.0.6

No need to revert it IMO. We can adapt these notations.

File: doc/specs/stdlib_math.md [hunk 1]
```diff
@@ -156,7 +156,7 @@ Returns a logarithmically spaced rank 1 array from [`base`^`start`, `base`^`end`
 
 #### Syntax
 
-`res = [[stdlib_math(module):logspace(interface)]] (start, end [, n [, base]])`
+`res = ` [[stdlib_math(module):logspace(interface)]] ` (start, end [, n [, base]] `)`
```
Comment (jvdp1) [inline doc/specs/stdlib_math.md:159]:
```suggestion
`res = ` [[stdlib_math(module):logspace(interface)]] ` (start, end [, n [, base]])`
```

File: doc/specs/stdlib_stats.md [hunk 2]
```diff
@@ -78,7 +78,7 @@ The scaling can be changed with the logical argument `corrected`. If `corrected`
 
 ### Syntax
 
-`result = [[stdlib_stats(module):cov(interface)]](array, dim [, mask [, corrected]])`
+`result = ` [[stdlib_stats(module):cov(interface)]] `(array, dim [, mask [, corrected]] `)`
```
Comment (jvdp1) [inline doc/specs/stdlib_stats.md:81]:
```suggestion
`result = ` [[stdlib_stats(module):cov(interface)]] `(array, dim [, mask [, corrected]])`
```

File: doc/specs/stdlib_stats.md [hunk 3]
```diff
@@ -248,9 +248,9 @@ The _k_-th order moment about `center` is defined as :
 
 ### Syntax
 
-`result = [[stdlib_stats(module):moment(interface)]](array, order [, center [, mask]])`
+`result = ` [[stdlib_stats(module):moment(interface)]] `(array, order [, center [, mask]] `)`
```
Comment (jvdp1) [inline doc/specs/stdlib_stats.md:251]:
```suggestion
`result = ` [[stdlib_stats(module):moment(interface)]] `(array, order [, center [, mask]])`
```

File: doc/specs/stdlib_stats.md [hunk 4]
```diff
@@ -248,9 +248,9 @@ The _k_-th order moment about `center` is defined as :
...
 
-`result = [[stdlib_stats(module):moment(interface)]](array, order, dim [, center [, mask]])`
+`result = ` [[stdlib_stats(module):moment(interface)]] `(array, order, dim [, center [, mask]] `)`
```
Comment (jvdp1) [inline doc/specs/stdlib_stats.md:253]:
```suggestion
`result = ` [[stdlib_stats(module):moment(interface)]] `(array, order, dim [, center [, mask]])`
```

File: doc/specs/stdlib_stats.md [hunk 5]
```diff
@@ -306,9 +306,9 @@ The use of the term `n-1` for scaling is called Bessel 's correction. The scalin
 
 ### Syntax
 
-`result = [[stdlib_stats(module):var(interface)]](array [, mask [, corrected]])`
+`result = ` [[stdlib_stats(module):var(interface)]] `(array [, mask [, corrected]] `)`
```
Comment (jvdp1) [inline doc/specs/stdlib_stats.md:309]:
```suggestion
`result = ` [[stdlib_stats(module):var(interface)]] `(array [, mask [, corrected]])`
```

File: doc/specs/stdlib_stats.md [hunk 6]
```diff
@@ -306,9 +306,9 @@ The use of the term `n-1` for scaling is called Bessel 's correction. The scalin
...
 
-`result = [[stdlib_stats(module):var(interface)]](array, dim [, mask [, corrected]])`
+`result = ` [[stdlib_stats(module):var(interface)]] `(array, dim [, mask [, corrected]] `)`
```
Comment (jvdp1) [inline doc/specs/stdlib_stats.md:311]:
```suggestion
`result = ` [[stdlib_stats(module):var(interface)]] `(array, dim [, mask [, corrected]])`
```

**Comment by jvdp1** [comment]
@perazz @ZedThree @gareth-nx :

Currently, this PR does:
* `Ford`: upgrade from `6.1.10` to `7.0.5`
* Fix all the specs to comply new `Ford` rules
* Exclude all LAPACK source files from the `Ford` processing (reason: 1) too many procedures => too long)

Processing the BLAS files does not significantly influence the time of the CI.

I propose that we merge this PR as is, even if the LAPACK files are excluded for `ford`. The main reason is to reduce the time of the CI of other PR.

**Comment by perazz** [comment]
LGTM @jvdp1

**Comment by perazz** [comment]
If there was a way to include the LAPACK interfaces (`stdlib_linalg_lapack.f90`), without FORD looking for all `module procedure` implementations, it would be great, but to the best of my knowledge, it is not possible.

**Review by gareth-nx — APPROVED** [review]
Thanks @jvdp1, this looks good. I agree it makes sense to exclude Lapack until we can address the doc build times.

**Comment by jvdp1** [comment]
Thank you for your reviews.

> If there was a way to include the LAPACK interfaces (`stdlib_linalg_lapack.f90`), without FORD looking for all `module procedure` implementations (`stdlib_linalg_lapack_*.f90`), it would be great, but to the best of my knowledge, it is not possible.

I tried several things similar to your proposition, but all failed! @ZedThree do you have maybe an idea? If not, I will wait a few more days and then merge it.

**Comment by ZedThree** [comment]
There is the (slightly unintuitive) `display: none` option that stops the rendering of any children of an entity -- except that it seems we still actually *process* the docstring, we just don't *display* it, which doesn't help.

It should be possible to fix this, unfortunately I'm at a conference basically all this week, so it might take me awhile to look into it. A quick hack does show some improvement:

```py
modified   ford/sourceform.py
@@ -612,6 +612,10 @@ class FortranBase:
     def filter_display(self, collection: Sequence) -> List:
         """Remove items from collection if they shouldn't be displayed"""
 
+        for obj in collection:
+            if not self._should_display(obj):
+                self.source_file._to_be_markdowned.remove(obj)
+
         return [obj for obj in collection if self._should_display(obj)]
```

but there still seems to be a pretty big slowdown.

**Comment by jvdp1** [comment]
Thank you @ZedThree for your comment. There is no rush for this issue IMO. I will merge this PR as is. 
You can look at this issue later if you want, and when you have time.

**Pull request merged**